File: cpp/src/core/preprocessing/baselines/rolling_ball.c
/* SPDX-License-Identifier: CECILL-2.1 */
/*
 * RollingBall — Kneen & Annegarn 1996 morphological baseline.
 *
 * Internal parity fixture: parity/python_generator/src/c4a_parity_pybaselines_ref/rolling_ball.py
 *
 * Implementation: O(n * w) scan with centred windows clipped at the array
 * boundaries. The pybaselines variant uses scipy.ndimage.{minimum,maximum}_filter
 * with `mode='nearest'`; for the parity reference we emit the equivalent
 * value-clipped-to-boundary behaviour (centred half-window R, window indices
 * [max(0, i-R), min(n-1, i+R)]).
 *
 * The optional final moving-average smoothing uses the same centred clipped
 * window pattern. `smooth_half_window == 0` disables smoothing — the output
 * is z directly.
 */

#include "rolling_ball.h"

#include <stdlib.h>

struct c4a_pp_rolling_ball_state_t {
    int32_t half_window;
    int32_t smooth_half_window;
};

c4a_pp_rolling_ball_state_t* c4a_pp_rolling_ball_state_new(int32_t half_window,
                                                             int32_t smooth_half_window) {
    if (half_window < 1 || smooth_half_window < 0) {
        return NULL;
    }
    c4a_pp_rolling_ball_state_t* s =
        (c4a_pp_rolling_ball_state_t*)malloc(sizeof(*s));
    if (s == NULL) return NULL;
    s->half_window        = half_window;
    s->smooth_half_window = smooth_half_window;
    return s;
}

void c4a_pp_rolling_ball_state_free(c4a_pp_rolling_ball_state_t* state) {
    free(state);
}

static void min_filter(const double* src, double* dst, int64_t n, int32_t w) {
    for (int64_t i = 0; i < n; ++i) {
        const int64_t lo = (i - w >= 0) ? (i - w) : 0;
        const int64_t hi = (i + w < n)  ? (i + w) : (n - 1);
        double m = src[lo];
        for (int64_t j = lo + 1; j <= hi; ++j) {
            if (src[j] < m) m = src[j];
        }
        dst[i] = m;
    }
}

static void max_filter(const double* src, double* dst, int64_t n, int32_t w) {
    for (int64_t i = 0; i < n; ++i) {
        const int64_t lo = (i - w >= 0) ? (i - w) : 0;
        const int64_t hi = (i + w < n)  ? (i + w) : (n - 1);
        double m = src[lo];
        for (int64_t j = lo + 1; j <= hi; ++j) {
            if (src[j] > m) m = src[j];
        }
        dst[i] = m;
    }
}

static void moving_avg(const double* src, double* dst, int64_t n, int32_t w) {
    for (int64_t i = 0; i < n; ++i) {
        const int64_t lo = (i - w >= 0) ? (i - w) : 0;
        const int64_t hi = (i + w < n)  ? (i + w) : (n - 1);
        double s = 0.0;
        for (int64_t j = lo; j <= hi; ++j) s += src[j];
        dst[i] = s / (double)(hi - lo + 1);
    }
}

c4a_status_t c4a_pp_rolling_ball_state_apply(
    const c4a_pp_rolling_ball_state_t* state,
    const double* X,
    int64_t rows, int64_t cols,
    double* out) {
    if (state == NULL || X == NULL || out == NULL) {
        return C4A_ERR_NULL_POINTER;
    }
    if (rows < 0 || cols < 0) {
        return C4A_ERR_INVALID_ARGUMENT;
    }
    if (rows == 0 || cols == 0) {
        return C4A_OK;
    }

    const int32_t W = state->half_window;
    const int32_t S = state->smooth_half_window;

    double* buf_a = (double*)malloc((size_t)cols * sizeof(double));
    double* buf_b = (double*)malloc((size_t)cols * sizeof(double));
    if (!buf_a || !buf_b) {
        free(buf_a); free(buf_b);
        return C4A_ERR_OUT_OF_MEMORY;
    }

    for (int64_t r = 0; r < rows; ++r) {
        const double* y = X + (size_t)r * (size_t)cols;
        /* erosion */
        min_filter(y, buf_a, cols, W);
        /* dilation */
        max_filter(buf_a, buf_b, cols, W);
        const double* z;
        if (S > 0) {
            moving_avg(buf_b, buf_a, cols, S);
            z = buf_a;
        } else {
            z = buf_b;
        }
        double* orow = out + (size_t)r * (size_t)cols;
        for (int64_t i = 0; i < cols; ++i) {
            orow[i] = y[i] - z[i];
        }
    }

    free(buf_a); free(buf_b);
    return C4A_OK;
}

File: parity/python_generator/src/n4m_parity_pybaselines_ref/rolling_ball.py
# SPDX-License-Identifier: CECILL-2.1
"""
RollingBall — Kneen & Annegarn 1996 morphological baseline.

For each row of length N:

  e[i] = min over [max(0, i - W), min(N-1, i + W)] of y
  z[i] = max over [max(0, i - W), min(N-1, i + W)] of e
  if smooth_half_window > 0:
      z = moving-average of z over a (2 * S + 1) clipped centred window
  return y - z

Edges use a centred window clipped to [0, N - 1] (no padding / reflection).
The C engine matches this reference bit-for-bit.

Defaults: half_window = 20, smooth_half_window = 0.
"""

from __future__ import annotations

import numpy as np


def _min_filter(y: np.ndarray, w: int) -> np.ndarray:
    n = y.size
    out = np.empty_like(y)
    for i in range(n):
        lo = max(0, i - w)
        hi = min(n - 1, i + w)
        out[i] = y[lo:hi + 1].min()
    return out


def _max_filter(y: np.ndarray, w: int) -> np.ndarray:
    n = y.size
    out = np.empty_like(y)
    for i in range(n):
        lo = max(0, i - w)
        hi = min(n - 1, i + w)
        out[i] = y[lo:hi + 1].max()
    return out


def _moving_avg(y: np.ndarray, w: int) -> np.ndarray:
    n = y.size
    out = np.empty_like(y)
    for i in range(n):
        lo = max(0, i - w)
        hi = min(n - 1, i + w)
        out[i] = y[lo:hi + 1].mean()
    return out


def rolling_ball(X: np.ndarray, *,
                  half_window: int = 20,
                  smooth_half_window: int = 0) -> np.ndarray:
    """
    Apply rolling-ball baseline correction row-wise to X.
    """
    if half_window < 1:
        raise ValueError("half_window must be >= 1")
    if smooth_half_window < 0:
        raise ValueError("smooth_half_window must be >= 0")

    X = np.ascontiguousarray(X, dtype=np.float64)
    if X.ndim != 2:
        raise ValueError("rolling_ball requires a 2-D input")
    rows, _cols = X.shape

    out = np.empty_like(X)
    for r in range(rows):
        y = X[r]
        z = _max_filter(_min_filter(y, half_window), half_window)
        if smooth_half_window > 0:
            z = _moving_avg(z, smooth_half_window)
        out[r] = y - z
    return out
